sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -515,3 +515,100 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
+       {"a":{"a":1,"b":2,"c":3}}
+  """)
+// scalastyle:on line.size.limit
+case class WithFields(children: Seq[Expression]) extends Unevaluable {
+
+  private lazy val structExpr = children.head
+  private lazy val structType = structExpr.dataType.asInstanceOf[StructType]
+  private lazy val (nameExprs, valExprs) = children.drop(1).grouped(2).map {
+    case Seq(name, value) => (name, value)
+  }.toList.unzip
+  private lazy val names = nameExprs.map(e => Option(e.eval()).map(_.toString).orNull)
+  private lazy val addOrReplaceExprs = names.zip(valExprs)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val expectedStructType = StructType(Nil).typeName
+    if (children.size % 2 == 0) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName expects an odd number of arguments.")
+    } else if (structExpr.dataType.typeName != expectedStructType) {
+      TypeCheckResult.TypeCheckFailure(
+        s"Only $expectedStructType is allowed to appear at first position, got: " +
+          s"${structExpr.dataType.typeName}.")
+    } else if (names.contains(null)) {
+      TypeCheckResult.TypeCheckFailure("Field name should not be null.")
+    } else if (nameExprs.exists(e => !(e.foldable && e.dataType == StringType))) {
+      TypeCheckResult.TypeCheckFailure(
+        s"Only foldable ${StringType.catalogString} expressions are allowed to appear at even " +
+          "position.")
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def dataType: StructType = {
+    val existingStructFields: Seq[(String, StructField)] = structType.fields.map { f =>
+      (f.name, f.copy(nullable = structExpr.nullable || f.nullable))
+    }
+    val addOrReplaceStructFields: Seq[(String, StructField)] = addOrReplaceExprs.map {
+      case (name, expr) => (name, StructField(name, expr.dataType, expr.nullable))
+    }
+    StructType(addOrReplace(existingStructFields, addOrReplaceStructFields).map(_._2))
+  }
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def nullable: Boolean = false
+
+  override def prettyName: String = "with_fields"
+
+  def toCreateNamedStruct: CreateNamedStruct = {
+    val existingExprs = structType.fieldNames.zipWithIndex.map {
+      case (name, i) => (name, GetStructField(structExpr, i, Some(name)))
+    }
+    val newExprs = addOrReplace(existingExprs, addOrReplaceExprs).flatMap {
+      case (name, valExpr) => Seq(Literal(name), valExpr)
+    }
+    CreateNamedStruct(newExprs)
+  }
+
+  private def addOrReplace[T](
+    existingFields: Seq[(String, T)],
+    addOrReplaceFields: Seq[(String, T)]): Seq[(String, T)] = {
+
+    val existingFieldNames = existingFields.map(_._1).toSet
+
+    addOrReplaceFields.foldLeft(existingFields) {
+      case (resultFields, newField @ (newFieldName, _)) =>
+        if (existingFieldNames(newFieldName)) {
+          resultFields.map {
+            case (name, _) if name == newFieldName => newField
```

Comment:
Do we consider case sensitive config?

Reply:
Good point, I didn't think about that. `CASE_SENSITIVE` config is not taken into account in the current implementation. I will fix this.